File: ews_PA_project/PA_project_java/src/progettoJava/Supermarket.java
package progettoJava;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

class Cart {
	ArrayList<Item> cartItems;
	boolean fidelityCard;

	public Cart(boolean fidalityCard) {
		this.cartItems = new ArrayList<>();
		this.fidelityCard = fidalityCard;
	}
	
	// bounded generics + varargs
	public <T extends Item> void addItems(T...newItems) {
		for(T item : newItems)
			this.cartItems.add(item);
	}
	
	public <T extends Item> void removeItem(T itemToRemove) {
		this.cartItems.remove(itemToRemove);
		System.out.println(itemToRemove.name + " rimosso");
	}
	
	public void printCart() {
		System.out.println("Contenuto carrello:");
		Supermarket.printItems(this.cartItems);
	}
}

// final class -> non può essere estesa
final class ShoppingVisitor implements Visitor {
	final double sundayDiscount = 0.9;
	
	public static int getDayNumber() {
		Date date = new Date();
	    return date.getDay();
	}
	
	@Override
	public double visit(ItemSoldInPieces item) {
		if(getDayNumber() == 7)
			return sundayDiscount * item.numberOfPieces * item.getUnitPrice();
		return item.numberOfPieces * item.getUnitPrice();
	}

	@Override
	public double visit(ItemSoldInWeight item) {
		if(getDayNumber() == 7)
			return sundayDiscount * item.weight * item.getUnitPrice();
		return item.weight * item.getUnitPrice();
	}
}

//class ShoppingVisitorDerived extends ShoppingVisitor{}   // non posso estendere

public class Supermarket {
	ArrayList<Item> supermarketItems;
	final static double fidelityDiscount = 0.85;   // final variable, non può essere cambiata
	// static variable, inizializzata solo ad inizio esecuzione, non legata al singolo oggetto
	
	// singleton (costruttore private, utile se costruzione oggetto onerosa)
	private Supermarket() {}
	private Supermarket(ArrayList<Item> suermarketItems) {
		this.supermarketItems = suermarketItems;
	}
	static private Supermarket supermarket = null;   // creo oggetto solo se richiesto esplicitamente

	static public Supermarket newSupermarket(ArrayList<Item> suermarketItems) {
		// return new Supermarket() non andrebbe bene perché restituirebbe ogni volta un
		// supermarket diverso
		if (supermarket == null)
			supermarket = new Supermarket(suermarketItems);
		return supermarket;
	}
	
	public static String getBill(Cart cart) {
		double result = 0;
		ShoppingVisitor shoppingVisitor = new ShoppingVisitor();
		for(Item item : cart.cartItems)
			result += item.accept(shoppingVisitor);
		if(cart.fidelityCard)
			result *= fidelityDiscount;
		DecimalFormat df = new DecimalFormat("0.00");
		return df.format(result);
	}
	
	// wildcard
	public static void printItems(List<? extends Item> itemsList) {
		for(int i = 0; i < itemsList.size(); i++)
			System.out.println(itemsList.get(i).toString());
	}
	
	public static <T> void printArrayItems(T[] itemsList) {
		System.out.println("Stampo prodotti nell'array:");
		for(int i = 0; i < itemsList.length; i++)
			System.out.println(itemsList[i].toString());
	}
	
	public <T extends Item> void removeItem(T itemToRemove) {
		this.supermarketItems.remove(itemToRemove);
		System.out.println(itemToRemove.name + " rimosso");
	}
	
	public static void main(String[] args) throws Throwable {
		// creazione di vari prodotti
		ItemSoldInPieces detersivo = new ItemSoldInPieces("detersivo", 2.32, 2);
		ItemSoldInPieces dentifricio = new ItemSoldInPieces("dentifricio", 2.80, 3);
		ItemSoldInPieces dopoBarba = new ItemSoldInPieces("dopobarba", 5.00, 1);
		ItemSoldInWeight cestoBanane = new ItemSoldInWeight("cesto banane", 3.53, 0.5);
		ItemSoldInWeight noci = new ItemSoldInWeight("noci", 8.90, 0.4);
		ItemSoldInWeight mele = new ItemSoldInWeight("mele", 0.88, 0.8);
		System.out.println("\n---------------- Prodotti creati ----------------\n");
		System.out.println(detersivo.toString());
		System.out.println(dentifricio.toString());
		System.out.println(dopoBarba.toString());
		System.out.println(cestoBanane.toString());
		System.out.println(noci.toString());
		System.out.println(mele.toString());
		
		System.out.println("\n---------------- Creo lista prodotti supermercato ----------------\n");
		ArrayList<Item> supermarketItemsList = new ArrayList<>();
		supermarketItemsList.add(detersivo);
		supermarketItemsList.add(dentifricio);
		supermarketItemsList.add(cestoBanane);
		supermarketItemsList.add(noci);
		supermarketItemsList.add(mele);
		
		System.out.println("---------------- Creo supermarket ----------------\n");
		// Supermarket supermarket = new Supermarket(); // non si può
		Supermarket supermarket = newSupermarket(supermarketItemsList); // creazione solo su richiesta		
		System.out.println("---------------- Stampo lista prodotti ----------------\n");
		printItems(supermarket.supermarketItems);   // sfrutto wildcard
		
		System.out.println("\n---------------- rimuovo prodotto dalla lista ----------------\n");
		supermarket.removeItem(mele);   
		printItems(supermarket.supermarketItems); 
		
		System.out.println("\n---------------- array di prodotti ----------------\n");
		Item[] itemsArray = new Item[3]; 
		itemsArray[0] = dentifricio;
		itemsArray[1] = noci;
		itemsArray[2] = detersivo;
		printArrayItems(itemsArray);
		System.out.println("\n---------------- array di prodotti riordinato ----------------\n");
		Item.orderListByPrice(itemsArray);
		printArrayItems(itemsArray);
		
		System.out.println("\n---------------- creo due carrelli, con e senza carta fedeltà ----------------\n");
		Cart cartFidelity = new Cart(true);
		Cart cartNotFidelity = new Cart(false);
		
		System.out.println("---------------- riempio i due carrelli ----------------\n");
		cartFidelity.addItems(detersivo);
		cartFidelity.addItems(dentifricio);
		cartFidelity.addItems(cestoBanane);
		cartNotFidelity.addItems(dentifricio, cestoBanane, noci);   // varargs
		
		System.out.println("---------------- stampo prodotti nel carrello ----------------\n");
		cartFidelity.printCart();
		System.out.println("\n---------------- rimuovo detersivo ----------------\n");
		cartFidelity.removeItem(detersivo);
		cartFidelity.printCart();

		System.out.println("\n---------------- stampo conto carrelli ----------------\n");
		System.out.println("cliente fedele: " + getBill(cartFidelity));
		System.out.println("cliente non fedele: " + getBill(cartFidelity));
		
	}
}

File: ews_PA_project/PA_project_cpp/src/Supermarket.cpp
#include <iostream>
#include <stdlib.h>
#include <stdio.h>
#include <vector>
#include <string.h>
#include "Item.h"
#include "Supermarket.h"
#include <ctime>
using namespace std;

Cart::Cart(bool fidelityCard){
	vector<ItemSoldInWeight> cartItemsSoldInPieces;
	vector<ItemSoldInWeight> cartItemsSoldInWeight;
	this->fidelityCard = fidelityCard;
}

Cart::~Cart(){}

void Cart::addItemSoldInPieces(ItemSoldInPieces newItem){
	cartItemsSoldInPieces.push_back(newItem);
}

void Cart::addItemSoldInWeight(ItemSoldInWeight newItem){
	cartItemSoldInWeight.push_back(newItem);
}

void Cart::removeItemSoldInPieces(ItemSoldInPieces itemToRemove) {
	int position = 0;
	vector<ItemSoldInPieces>::iterator it;
	for (unsigned int i = 0; i < cartItemsSoldInPieces.size(); i++) {
		if (cartItemsSoldInPieces[i].equalsTo(itemToRemove)) {
			position = i;
			break;
		}
	}
	it = cartItemsSoldInPieces.begin() + position;
	cartItemsSoldInPieces.erase(it);
}

void Cart::removeItemSoldInWeight(ItemSoldInWeight itemToRemove) {
	int position = 0;
	vector<ItemSoldInWeight>::iterator it;
	for (unsigned int i = 0; i < cartItemSoldInWeight.size(); i++) {
		if (cartItemSoldInWeight[i].equalsTo(itemToRemove)) {
			position = i;
			break;
		}
	}
	it = cartItemSoldInWeight.begin() + position;
	cartItemSoldInWeight.erase(it);
}

void Cart::printCart() {
	printf("Prodotti nel carrello:\n");
	printf("venduti a pezzi:\n");
	for (unsigned int i = 0; i < cartItemsSoldInPieces.size(); i++) {
		printf("%s\n", cartItemsSoldInPieces[i].toString());
	}
	printf("venduti a peso:\n");
	for (unsigned int i = 0; i < cartItemSoldInWeight.size(); i++) {
		printf("%s\n", cartItemSoldInWeight[i].toString());
	}
}


Supermarket::Supermarket(vector<ItemSoldInPieces> supermarketItemSoldInPieces,
		vector<ItemSoldInWeight> supermarketItemSoldInWeight) {
	this->supermarketItemSoldInPieces = supermarketItemSoldInPieces;
	this->supermarketItemSoldInWeight = supermarketItemSoldInWeight;
}

void Supermarket::printItems() {
	printf("Prodotti venduti a pezzi:\n");
	for (unsigned int i = 0; i < supermarketItemSoldInPieces.size(); i++) {
		printf("%s\n", supermarketItemSoldInPieces[i].toString());
	}
	printf("Prodotti venduti a peso:\n");
	for (unsigned int i = 0; i < supermarketItemSoldInWeight.size(); i++) {
		printf("%s\n", supermarketItemSoldInWeight[i].toString());
	}
}

void Supermarket::addItemSoldInPieces(ItemSoldInPieces newItem){
	supermarketItemSoldInPieces.push_back(newItem);
}
void Supermarket::addItemSoldInWeight(ItemSoldInWeight newItem){
	supermarketItemSoldInWeight.push_back(newItem);
}

void Supermarket::removeItemSoldInPieces(ItemSoldInPieces itemToRemove) {
	int position = 0;
	vector<ItemSoldInPieces>::iterator it;
	for (unsigned int i = 0; i < supermarketItemSoldInPieces.size(); i++) {
		if(supermarketItemSoldInPieces[i].equalsTo(itemToRemove)){
			position = i;
			break;
		}
	}
	it = supermarketItemSoldInPieces.begin() + position;
	supermarketItemSoldInPieces.erase(it);
}

void Supermarket::removeItemSoldInWeight(ItemSoldInWeight itemToRemove) {
	int position = 0;
	vector<ItemSoldInWeight>::iterator it;
	for (unsigned int i = 0; i < supermarketItemSoldInWeight.size(); i++) {
		if(supermarketItemSoldInWeight[i].equalsTo(itemToRemove)){
			position = i;
			break;
		}
	}
	it = supermarketItemSoldInWeight.begin() + position;
	supermarketItemSoldInWeight.erase(it);
}

double Supermarket::getBill(Cart cart) {
	double result = 0;
	time_t now = time(0);  // ora corrente
	tm *localTime = localtime(&now);   // conversione a calendario
	int sunday = localTime->tm_wday;   // giorni passati dall'ultima domenica
	for (unsigned int i = 0; i < cart.cartItemsSoldInPieces.size(); i++)
		result += cart.cartItemsSoldInPieces[i].getItemPrice();
	for (unsigned int i = 0; i < cart.cartItemSoldInWeight.size(); i++)
		result += cart.cartItemSoldInWeight[i].getItemPrice();
	if (sunday == 0)
		result *= fidelityDiscount;
	if(cart.fidelityCard)
		result *= fidelityDiscount;
	return result;
}


int main() {
	// creazione di vari prodotti
	char det[] = "detersivo";
	char dent[] = "dentifricio";
	char db[] = "dopobarba";
	char cb[] = "cesto banane";
	char no[] = "noci";
	char me[] = "mele";
	ItemSoldInPieces detersivo(det, 2.32, 2);
	ItemSoldInPieces dentifricio(dent, 2.80, 3);
	ItemSoldInPieces dopoBarba(db, 5.00, 1);
	ItemSoldInWeight cestoBanane(cb, 3.53, 0.5);
	ItemSoldInWeight noci(no, 8.90, 0.4);
	ItemSoldInWeight mele(me, 0.88, 0.8);
	printf("\n---------------- Prodotti creati ----------------\n");
	printf("%s\n", detersivo.toString());
	printf("%s\n", dentifricio.toString());
	printf("%s\n", dopoBarba.toString());
	printf("%s\n", cestoBanane.toString());
	printf("%s\n", noci.toString());
	printf("%s\n", mele.toString());

	printf("---------------- Creo liste di prodotti ----------------\n");
	vector<ItemSoldInPieces> supermarketItemSoldInPieces;
	supermarketItemSoldInPieces.push_back(detersivo);
	supermarketItemSoldInPieces.push_back(dentifricio);
	vector<ItemSoldInWeight> supermarketItemSoldInWeight;
	supermarketItemSoldInWeight.push_back(noci);
	supermarketItemSoldInWeight.push_back(mele);

	printf("---------------- creo supermarket ----------------\n");
	Supermarket conad(supermarketItemSoldInPieces, supermarketItemSoldInWeight);
	printf("---------------- aggiungo dopobarba----------------\n");
	conad.addItemSoldInPieces(dopoBarba);
	conad.printItems();
	printf("\n---------------- rimuovo dopobarba ----------------\n");
	conad.removeItemSoldInPieces(dopoBarba);
	conad.printItems();

	printf("\n---------------- creo due carrelli, con e senza carta fedeltà ----------------\n");
	Cart cartFidelity (true);
	Cart cartNotFidelity (false);

	printf("---------------- riempio i due carrelli ----------------\n");
	cartFidelity.addItemSoldInPieces(detersivo);
	cartFidelity.addItemSoldInPieces(dentifricio);
	cartFidelity.addItemSoldInWeight(cestoBanane);
	cartFidelity.addItemSoldInWeight(noci);
	cartNotFidelity.addItemSoldInPieces(dentifricio);
	cartNotFidelity.addItemSoldInWeight(cestoBanane);
	cartNotFidelity.addItemSoldInWeight(noci);

	printf("---------------- stampo prodotti nel carrello ----------------\n");
	cartFidelity.printCart();
	printf("\n---------------- rimuovo detersivo ----------------\n");
	cartFidelity.removeItemSoldInPieces(detersivo);
	cartFidelity.printCart();

	printf("\n---------------- stampo conto carrelli ----------------\n");
	printf("cliente fedele: %.2f€\n", conad.getBill(cartFidelity));
	printf("cliente non fedele: %.2f€\n", conad.getBill(cartNotFidelity));


	return 0;
}



